src/edu/csus/ecs/pc2/ui/TestResultsPane.java

```diff
@@ -1013,6 +1015,63 @@ private JPanel getResultsPaneButtonPanel() {
         }
         return resultsPaneButtonPanel;
     }
+
+    /**
+     * refreshes the resultspanebuttonpanel
+     * includes/removes execute all button based on stop-on-first-failure
+     */
+    private void updateResultsPaneButtonPanel() {
```

Comment:
I kind of figured out a way to fix this however one small issue that arises is that the Execute All button now comes up in the end after the Close button. So far, I'm unable to figure out how to change the ordering of the buttons. I'm still looking into it. Basically, the issue is that there is a JPanel in FlowLayout with a bunch of components, how to now add some new components in between some existing components.

![image](https://github.com/pc2ccs/pc2v9/assets/24360328/ab303418-ba0e-4eaf-8d08-92cbb1c0aa5f)

Reply:
One solution, admittedly not necessarily ideal, would be to create two sub-panels in the "Button Panel":  one to hold the "permanent" buttons (the checkbox, plus the `Select All`, `Unselect All`, `Compare Selected`, and `Execute All` buttons), and a second panel to hold (just) the `Close` button.  Then you could add/remove the `Execute All` button from the "left sub-panel" and the "right sub-panel" containing (just) the `Close` button would always appear to the right of whatever was in the "left sub-panel".  This however would certainly add some additional (perhaps undesirable) complexity to the GUI.  Another solution might be to use a `GridBagLayout` for the button panel -- but if you've never fooled with those, fair warning: they are pretty complicated (but, very powerful for controlling layouts).

A better solution of course would be if there's a way to manage the insertion point of a component in a JPanel with a flow layout -- but I'm not sure that's possible without rebuilding the entire panel (which as I pointed out isn't really desirable -- at least, there shouldn't be two different places where it's done...)

If none of the above helps, my own feeling is that it's better to have proper code organization, at the expense of having the `Execute All` button show up on the far left (others may feel differently...)  Overall, I would investigate the "two sub-panels" idea first; it might be relatively clean and easy to do.

Reply:
Updated with the "two sub-panels" approach

Reply:
I'm a little late to the party, but I have been doing some testing on this branch.
Instead of a left/right panel, why not just always add the execute all button, but just set its _visiblity_:
`getExecuteAllButton().setVisible(isExecuteAllButtonVisible) `?  Basically the button will always "be there", just not visible for the cases we don't want to see it.  I suppose you could do the same for the spacing strut too.  I quickly tried this and it seems to work.  Basically, as a test, I just added:
            
```
            //add space
            getResultsPaneButtonPanelLeft().add(getExecuteAllHorizontalStrut());
            
            // add a control button to execute all test cases
            getResultsPaneButtonPanelLeft().add(getExecuteAllButton());
            
            getExecuteAllButton().setVisible(false);

```
to your "`getResultsPaneButtonPanelLeft()`" method.  And changed `updateResultsPaneButtonPanelLeft()` to:

```
    private void updateResultsPaneButtonPanelLeft() {

        if (currentProblem != null && currentProblem.isStopOnFirstFailedTestCase()) {
            isExecuteAllButtonVisible = true;
        }
        else {

            isExecuteAllButtonVisible = false;
        }
        getExecuteAllButton().setVisible(isExecuteAllButtonVisible);
    }

```
I think this would work as it was before you added the left and right panels.  Or, did you try this and find it didn't work?  It seems cleaner, assuming of course it works without the left/right panels.

Reply:
Updated as per suggestion. Earlier I had tried this approach, must have done something wrong as I was seeing a large space when the Execute All button set to be invisible.